### DISCUSSION TITLE
Problem with NRF24L01+ Transceiver, OSError Hardware not responding

### DISCUSSION THREAD

**Discussion by castus** [micropython/micropython#9776] [Hardware & Peripherals]
Hi,

I have a problem. I'm new to Pico and IoT. I wanted to test NRF24L01+ transceiver but I'm unable to run it. I'm constantly receiving an error 

```
OSError: nRF24L01+ Hardware not responding
```

I'm using Pico H 2021 with connections like that:
CE on 17
CSN on 14
SCK on 6
MISO on 4
MOSI on 7
SPI 0

and script: [nrf24l01.py](https://github.com/micropython/micropython/blob/d84c6ef0e8dd363881d80b2d8fb03447cc349830/drivers/nrf24l01/nrf24l01.py) and test script [nrf24l01test.py](https://github.com/micropython/micropython/blob/d84c6ef0e8dd363881d80b2d8fb03447cc349830/drivers/nrf24l01/nrf24l01test.py) with changed configuration to:

```
if usys.platform == "pyboard":
    cfg = {"spi": 2, "miso": "Y7", "mosi": "Y8", "sck": "Y6", "csn": "Y5", "ce": "Y4"}
elif usys.platform == "esp8266":  # Hardware SPI
    cfg = {"spi": 1, "miso": 12, "mosi": 13, "sck": 14, "csn": 4, "ce": 5}
elif usys.platform == "esp32":  # Software SPI
    cfg = {"spi": -1, "miso": 32, "mosi": 33, "sck": 25, "csn": 26, "ce": 27}
elif usys.platform == "rp2":  # PI PICO
    cfg = {"spi": 0, "miso": 4, "mosi": 7, "sck": 6, "csn": 14, "ce": 17}
else:
    raise ValueError("Unsupported platform {}".format(usys.platform))
```

I triple-checked connections and everything is in place but it's not working.

Could you help me to find out what's going on?

I really appreciate any help you can provide.

**Comment by peterhinch** [comment]
You are probably the first to try NRF24L01 with RP2. I think the problem is with [this line](https://github.com/micropython/micropython/blob/d84c6ef0e8dd363881d80b2d8fb03447cc349830/drivers/nrf24l01/nrf24l01test.py#L38). RP2 seems to require SPI to be initialised with named args for `sck` `mosi` and `miso`.

The quick fix is to change the SPI number to -1 which uses that form of initialisation.

**Comment by castus** [comment]
Unfortunately, this doesn't work. It gives me an error that there is no SPI -1. I gave up on this. It's tough to communicate between those modules even though there are many videos on YT. Thanks anyway @peterhinch

**Reply by jimmo** [reply]
Instead of `SPI(-1, ...pins)`, use `SoftSPI(...pins)`.

i.e. in nrf24l01test.py, add your pin config to the dict on line 23, then change line 35 to say `spi = SoftSPI(sck=...)` rather than `spi = SPI(-1, sck=...)`.

**Reply by peterhinch** [reply]
@jimmo Of course you are right. This means that [nrf24l01_test.py](https://github.com/micropython/micropython/blob/d84c6ef0e8dd363881d80b2d8fb03447cc349830/drivers/nrf24l01/nrf24l01test.py) is out of date as I copied the -1 value from the ESP32 case. I'll submit a PR to fix this for ESP32 and RP2.

**Reply by peterhinch** [reply]
I have raised https://github.com/micropython/micropython-lib/pull/568 which addresses this and other issues with the test script.
